iOS/Flows/Conversation/Cells/MessagesTimeMachineCollectionViewLayout.swift

```diff
@@ -0,0 +1,185 @@
+/// as the items move along the z axis.
+class MessagesTimeMachineCollectionViewLayout: TimeMachineCollectionViewLayout {
+
+    override class var layoutAttributesClass: AnyClass {
+        return ConversationMessageCellLayoutAttributes.self
+    }
+
+    // MARK: - Layout Configuration
+
+    /// How bright the background of the frontmost item is. 0 is black, 1 is full brightness.
+    var frontmostBrightness: CGFloat = 0.89
+    /// How bright the background of the backmost item is. This is based off of the frontmost item brightness.
+    var backmostBrightness: CGFloat {
+        return self.frontmostBrightness - CGFloat(self.stackDepth+1)*0.05
+    }
+
+    override func layoutAttributesForItemAt(indexPath: IndexPath,
+                                            withNormalizedZOffset normalizedZOffset: CGFloat) -> UICollectionViewLayoutAttributes? {
+
+        let attributes = super.layoutAttributesForItemAt(indexPath: indexPath,
+                                                         withNormalizedZOffset: normalizedZOffset)
+
+        guard let attributes = attributes as? ConversationMessageCellLayoutAttributes else {
+            return attributes
+        }
+
+
+        var backgroundBrightness: CGFloat
+        if normalizedZOffset < 0 {
+            // Dark the item as it moves away
+            backgroundBrightness = lerp(abs(normalizedZOffset),
+                                        start: self.frontmostBrightness,
+                                        end: self.backmostBrightness)
+        } else {
+            // Items should be at full when at the front of the stack.
+            backgroundBrightness = self.frontmostBrightness
+        }
+
+        let detailAlpha = 1 - abs(normalizedZOffset) / 0.2
+        let textViewAlpha = 1 - abs(normalizedZOffset) / 0.8
+
+        // The most recent visible item should be white.
+        if let itemFocusPosition = self.itemFocusPositions[indexPath] {
+            let normalizedFocusDistance = abs(itemFocusPosition - self.zPosition)/self.itemHeight.half
+
+            backgroundBrightness += lerpClamped(normalizedFocusDistance,
+                                                start: 1-self.frontmostBrightness,
+                                                end: 0)
+        }
+
+        // If there is no message to display for this index path, don't show the cell.
+        if self.dataSource?.getTimeMachineItem(forItemAt: indexPath) == nil {
+            attributes.alpha = 0
+        }
+
+        attributes.backgroundColor = .white
+        attributes.brightness = backgroundBrightness
+        attributes.shouldShowTail = indexPath.section == 0
+        attributes.bubbleTailOrientation = indexPath.section == 0 ? .up : .down
+        attributes.detailAlpha = detailAlpha
+        attributes.messageContentAlpha = self.isShowingDropZone && indexPath.section == 1 ? 0.0 : textViewAlpha
```

Comment:
Only the message cell specific attributes are calculated here. The base class handles the basics, like size, alpha and position.